File: loomoroutines-common/src/main/java/dev/reformator/loomoroutines/common/internal/Registry.java
package dev.reformator.loomoroutines.common.internal;

import dev.reformator.loomoroutines.common.RunningCoroutine;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class Registry {
    private Registry() { }

    public static final @NotNull Scoped<List<RunningCoroutine<?>>> runningCoroutinesScoped = new ThreadLocalScoped<>();

    public static final @NotNull CoroutineFactory coroutineFactory = LoomSuspendedCoroutine.factory;
}

File: loomoroutines-common/src/main/kotlin/internal/registry.kt
package dev.reformator.loomoroutines.common.internal

import dev.reformator.loomoroutines.common.Action
import dev.reformator.loomoroutines.common.Generator
import dev.reformator.loomoroutines.common.RunningCoroutine

val runningCoroutinesScoped: Scoped<MutableList<RunningCoroutine<*>>> = ThreadLocalScoped()

val coroutineFactory: CoroutineFactory = LoomCoroutineFactory

private val newRunningCoroutinesListGenerator = Generator<MutableList<RunningCoroutine<*>>> { mutableListOf() }

fun performInRunningCoroutinesScope(action: Action) {
    runningCoroutinesScoped.performReusable(newRunningCoroutinesListGenerator, action)
}

val runningCoroutinesInternal: MutableList<RunningCoroutine<*>>
    get() = runningCoroutinesScoped.get()
